neutrons/SNAPRed
# Ewm 4058 - Automate State Initialization

## Description of work
This task focused on refining the initialization state workflow, introducing an automated verification feature that activates each time a run number is submitted. Now, SNAPRed efficiently determines if the provided run number is linked to an existing state. If a match is found, the process seamlessly continues as planned. Conversely, if no associated state is detected, the system initiates an alternative workflow, offering users the option to initialize the state, thereby ensuring a smoother and more intuitive user experience.



## Explanation of work
The introduction of this feature presents a unique challenge in terms of complexity, given the intricate and somewhat enigmatic aspects of SNAPRed's operational flow. However, I will endeavor to demystify how this enhancement works.

Embedded within the Calibration and Normalization workflows are input fields where users can enter a run number for assessment. After completing the required fields, users proceed by clicking the Continue button. The logic governing this action is encapsulated within the WorkflowPresenter, specifically in the _handleContinueButtonClicked method. This method now includes a refined process, _handleComplications, engineered to accommodate a newly defined error type, RecoverableError.

On the backend, particularly within the error management directory, we've introduced a bespoke exception named RecoverableError. This is strategically employed alongside the ExecutionHandler in the LocalDataService to intercept exceptions triggered by the readCalibrationState method. Furthermore, adjustments to SNAPResponse have been made to facilitate the detection of such exceptions, thereby triggering the initiation of the state initialization workflow. This comprehensive overhaul not only integrates this advanced feature into the existing framework but also enhances the interface and workflow mechanisms for a smoother user interaction.


## To test

To test this feature, open SNAPRed GUI and complete the DiffractionCalibration Workflow form. Values used:
```python
Run Number: 46581
Convergence Threshold: 0.1
Peak Intensity Threshold: 0.0001
Bins Across Peak Width: 10
Sample: Any
Grouping File: Any
Peak Function: Any
``` 

This should result in a pop-up that prompts the user if they want to initialize the state. Please test this feature to completion. Test with other non-initialized run numbers.

## Link to EWM item


[EWM# 4058](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=4058)
